include/deal.II/distributed/cell_data_transfer.h

```diff
@@ -0,0 +1,357 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_distributed_cell_data_transfer_h
+#define dealii_distributed_cell_data_transfer_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/smartpointer.h>
+
+#include <deal.II/distributed/tria.h>
+
+#include <boost/range/iterator_range.hpp>
+
+#include <functional>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace distributed
+  {
+    /**
+     * Transfer data that is associated with each active cell (like error
+     * indicators) while refining and/or coarsening a distributed triangulation
+     * and handle the necessary communication.
+     *
+     * This class therefore does for cell-related information what
+     * parallel::distributed::SolutionTransfer does for the values of degrees of
+     * freedom defined on a parallel::distributed::Triangulation.
+     *
+     * This class has been designed to operate on any kind of datatype that is
+     * serializable. A non-distributed Vector container has to be provided,
```

Comment:
Could you please add a test which transfers a pair if integers (something simple like 1 cell -> 4 cells)?
I see that you assume existence of `VectorType::value_type` so it should work nicely with `std::vector<std::pair<int,int>>`.

Reply:
This would be possible, but from my understanding, we would have to supply another explicit template specialization for it (as in `cell_data_transfer.inst.in`).

Do we want to do that? Or is there a way around it that I couldn't come up with?

Reply:
> Do we want to do that?

no, we don't. We should move all the definitions to the `.templates.h`, instantiate for things we find most useful (like `Vector<double>`) but still let users include both headers (i.e. `cell_data_transfer.h` and `cell_data_transfer.templates.h`) to be used with custom data types and containers so long as they provide things like `value_type`.

p.s. @bangerth please correct me if I am wrong in the interpretation of the `.templates.h` and `.h` in deal.II.